### PR TITLE
[SPARK-10500][SPARKR] sparkr.zip cannot be created if /R/lib is unwritable

### PR DESCRIPTION
**Pull request by sun-rui**
The basic idea is that:
The archive of the SparkR package itself, that is sparkr.zip, is created during build process and is contained in the Spark binary distribution. No change to it after the distribution is installed as the directory it resides ($SPARK_HOME/R/lib) may not be writable.

When there is R source code contained in jars or Spark packages specified with "--jars" or "--packages" command line option, a temporary directory is created by calling Utils.createTempDir() where the R packages built from the R source code will be installed. The temporary directory is writable, and won't interfere with each other when there are multiple SparkR sessions, and will be deleted when this SparkR session ends. The R binary packages installed in the temporary directory then are packed into an archive named rpkg.zip.

sparkr.zip and rpkg.zip are distributed to the cluster in YARN modes.

The distribution of rpkg.zip in Standalone modes is not supported in this PR, and will be address in another PR.

Various R files are updated to accept multiple lib paths (one is for SparkR package, the other is for other R packages)  so that these package can be accessed in R.

### CODE REVIEW

**Comment by shivaram** [comment]
cc @brkyvz @andrewor14

**Comment by shivaram** [comment]
@sun-rui Thanks for the PR. It looks like this changes a lot of parts of the code though, so we'll need to  review it carefully. 

Are there any alternate designs you have in mind that might not change our JVM to R protocol etc. ?

**Comment by sun-rui** [comment]
@shivaram, there are two changes to JVM to R protocol:
1. Env variable format used to convey SparkR package path to R. Previously a single path for the SparkR package is conveyed, in this PR, a comma-separated path list is conveyed. The first element is the path for SparkR package, the other is for additional R packages specified via spark-submit command line options.
2. Within sparkR.init(), after launching a JVM backend, a path for additional R packages is passed from the JVM backend to R. The path is then added into .libPaths() so that the additional R packages can be loaded within R environment.

So the basic change is that we have separate path for the SparkR package itself and additional R packages. An alternative design is:
When there are additional R packages, zip them and the SparkR packge into an archive in a temporary directory. Then we can distribute only one file to cluster and the change No.1 is not needed.
However, I think this alternative design has advantages:
a. Each time there are additional R packages, the SparkR packge has to be re-zipped.
b. With standalone mode cluster, actually only additional R packages need to be distributed. But in this design, the SparkR packge will be redistributed together, which is not necessary, and waste of network traffic.

So I think the current PR is more efficient and potentially more flexible.

No matter which design, change No.2 is necessary is for SparkR shell, because there needs to be a way for SparkR shell to access the additional R packages (the location of these packages is available after the R shell launched.)

**Comment by felixcheung** [comment]
Why not keep the existing `SPARKR_PACKAGE_DIR` and add another one for additional packages?

**Comment by sun-rui** [comment]
I could. But I think it is simpler to pass a list path using the existing env var instead of adding a new env var.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -362,22 +363,41 @@ object SparkSubmit {
...
+        }
+
+        val rPackageURI = Utils.resolveURI(rPackageFile.getAbsolutePath).toString
+        // Assigns a symbol link name "rpkg" to the shipped package.
+        args.archives = mergeFileLists(args.archives, rPackageURI + "#rpkg")
+      }
+    }
+
+    // TODO: Support distributing R packages with standalone cluster
+    if (args.isR && clusterManager == STANDALONE && !RUtils.rPackages.isEmpty) {
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: what about MESOS?

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 2]
```diff
@@ -51,15 +59,30 @@ private[spark] object RUtils {
...
+      val sparkRPkgPath = new File("sparkr").getAbsolutePath
+      val rPkgPath = new File("rpkg")
+      if (rPkgPath.exists()) {
+        Seq(sparkRPkgPath, rPkgPath.getAbsolutePath)
+      } else {
+        Seq(sparkRPkgPath)
+      }
     } else {
       // Otherwise, assume the package is local
       // TODO: support this for Mesos
```
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: I think we need to check and fail Mesos explicitly..

Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: Add check for running SparkR with Mesos, exit SparkSubmit in this case.
Submitted https://issues.apache.org/jira/browse/SPARK-11524 - "Support SparkR with Mesos cluster"

**Comment by sun-rui** [comment]
submitted https://issues.apache.org/jira/browse/SPARK-11525: Support spark packages containing R source code in Standalone mode

**Comment by shivaram** [comment]
@sun-rui Do R packages not work in standalone mode right now or is this something we are breaking as a result of this change ?

Ping @brkyvz to take a look at this PR

**Comment by sun-rui** [comment]
@shivaram, R packages not work in standalone mode is an existing bug, not a result of this PR.

**Comment by sun-rui** [comment]
I will test this PR on YARN cluster when I am free.

File: R/install-dev.bat [hunk 3]
```diff
@@ -25,3 +25,9 @@ set SPARK_HOME=%~dp0..
 MKDIR %SPARK_HOME%\R\lib
 
 R.exe CMD INSTALL --library="%SPARK_HOME%\R\lib"  %SPARK_HOME%\R\pkg\
+
+rem Zip the SparkR package so that it can be distributed to worker nodes on YARN
+pushd %SPARK_HOME%\R\lib
```
Comment (brkyvz) [inline R/install-dev.bat]: If `R/lib` is not writable, how are you going to write it here?

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 4]
```diff
@@ -23,6 +23,10 @@ import java.util.Arrays
 import org.apache.spark.{SparkEnv, SparkException}
 
 private[spark] object RUtils {
+  // Local path where R binary packages built from R source code contained in the spark
+  // packages specified with "--packages" or "--jars" command line option reside.
+  var rPackages: Option[String] = None
```
Comment (brkyvz) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: I find this hard to maintain. Where `RUtils.rPackages` is called would change from deployment type, right? The path will change from cluster mode to client, right?

File: R/install-dev.bat (hunk 3 above)
Comment (sun-rui) [inline R/install-dev.bat]: This happens in build process, not runtime.

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala (hunk 4 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: This path is set only on localhost, won't change ever. This path is similar and a companion one to the local SparkR package path (<SPARK_HOME>/R/lib), will only be retrieved when running SparkR in client modes.

**Comment by shivaram** [comment]
@brkyvz could you take another look ? It'll be good to get this PR in for 1.6. I'll also try to do a pass tonight

**Comment by sun-rui** [comment]
I tested this PR on yarn-client/yarn-cluster with/without R source packages. All work fine.

I may investigate later adding test cases for YARN mode.

**Comment by brkyvz** [comment]
@sun-rui I tested this package locally. I failed to use an R Spark Package. The problem is that if that package depends on SparkR, it can't find it during R CMD INSTALL since they don't reside in the same folder anymore.

**Comment by brkyvz** [comment]
@sun-rui With this patch, you should be able to enable this test:
https://github.com/apache/spark/blob/27524a3a9ccee6fbe56149180ebfb3f74e0957e7/core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala#L372
This way you can make sure that you don't break any existing behavior.
Regarding libraries not being able to find SparkR, one option would be to copy `SparkR` under `R/lib` to the temporary folder you create for building all other packages. Then you can zip everything except SparkR in that folder for Yarn support, and continue

**Comment by sun-rui** [comment]
Thanks to @brkyvz for catching this issue. I think a better solution is as documented at https://stat.ethz.ch/R-manual/R-devel/library/base/html/libPaths.html:
The library search path is initialized at startup from the environment variable R_LIBS (which should be a colon-separated list of directories at which R library trees are rooted) followed by those in environment variable R_LIBS_USER. Only directories which exist at the time will be included.

Before installing R sources, R_LIBS can be set to point to <SPARK_HOME>/R/lib.

**Comment by felixcheung** [comment]
I don't fully understand the issue, but why we have to use `.libPaths` and not `library(... lib.loc= )`?

**Comment by sun-rui** [comment]
@felixcheung, "R CMD INSTALL" is used to install R packages. Now if any R package depends on SparkR package, the installation will fail as the SparkR package path is not put into the R search library paths.

**Comment by sun-rui** [comment]
@brkyvz, setting env variable like 'R_LIBS' before running "R CMD INSALL" can fix this. But I will try to be consistent with the existing style that use "R_PROFILE_USER" env variable to specify an R script that will be automatically executed upon R is started. That is, using <SPARK_HOME>/R/lib/SparkR/profile/general.R to add the SparkR binary package path into .libPaths().

**Comment by brkyvz** [comment]
@sun-rui Tested this and it works like a charm. Thanks for the update. Now we're also able to enable the test thanks to you! I'll make a quick pass over the style (some comments would be helpful at points) and will leave the final vote to @shivaram

File: R/pkg/inst/worker/daemon.R [hunk 5]
```diff
@@ -18,10 +18,11 @@
 # Worker daemon
 
 rLibDir <- Sys.getenv("SPARKR_RLIBDIR")
-script <- paste(rLibDir, "SparkR/worker/worker.R", sep = "/")
+dirs <- strsplit(rLibDir, ",")[[1]]
+script <- paste(dirs[[1]], "SparkR/worker/worker.R", sep = "/")
```
Comment (brkyvz) [inline R/pkg/inst/worker/daemon.R]: didn't you already get `[[1]]` above?

Comment (brkyvz) [inline R/pkg/inst/worker/daemon.R]: also, shouldn't you use the `File.separator` equivalent as the `sep`? What if this is being run on Windows?

File: R/pkg/inst/profile/general.R [hunk 6]
```diff
@@ -17,6 +17,7 @@
 
 .First <- function() {
   packageDir <- Sys.getenv("SPARKR_PACKAGE_DIR")
-  .libPaths(c(packageDir, .libPaths()))
+  dirs <- strsplit(packageDir, ",")[[1]]
```
Comment (brkyvz) [inline R/pkg/inst/profile/general.R]: also comment what exists in `[[1]]`. Is it SparkR directory or packages?

File: core/src/main/scala/org/apache/spark/api/r/RUtils.scala [hunk 7]
```diff
@@ -34,11 +38,15 @@ private[spark] object RUtils {
   }
 
   /**
-   * Get the SparkR package path in various deployment modes.
+   * Get the list of paths for R packages in various deployment modes, of which the first
+   * path is for the SparkR package itself and the second if exists is for R packages
```
Comment (brkyvz) [inline core/src/main/scala/org/apache/spark/api/r/RUtils.scala]: `path is for the SparkR package itself. The second value is for R packages built as part of Spark Packages, if any exist.  Spark Packages can be provided through the --packages or --jars command line options` or something like that. Very long sentences are hard to read :)

File: R/pkg/inst/worker/daemon.R (hunk 5 above)
Comment (sun-rui) [inline R/pkg/inst/worker/daemon.R]: this is about the result of strsplit, which is a little bit strange. for example
```> strsplit("a,b,c", ",")
[[1]]
[1] "a" "b" "c"
```

For File.separator, will fix it

File: R/pkg/inst/profile/general.R (hunk 6 above)
Comment (sun-rui) [inline R/pkg/inst/profile/general.R]: same as above

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by sun-rui** [comment]
@brkyvz, the infinite loop if the environment is not cleared is because the inheritence of the R_PROFILE_USER env var set for launching SparkR, which causes recursive execution of sparkR.init(). By resetting R_PROFILE_USER to point to $SPARK_HOME/R/lib/profile/general.R, the SparkR lib path is added into the R lib search paths, as well as elimination of the infinite loop. "one stone, two birds":)

**Comment by sun-rui** [comment]
don't know what's going error with Jenkins

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
Thanks @sun-rui and @brkyvz -- Sorry for the delay in getting to this. LGTM. Merging this to master and branch-1.6

**Comment by sun-rui** [comment]
This is probably because when the test is being executed, SparkR library is not available under $SPARK_HOME/R/lib. Does the build has "-Psparkr" specified?

**Comment by shivaram** [comment]
Great catch @sun-rui -- I think it doesn't have `-Psparkr` -- Looking at the config xml file I see the command being run as 
```
 build/mvn --force -DzincPort=$ZINC_PORT -DskipTests -P${HADOOP_PROFILE} -Pyarn -Phive -Phive-thriftserver -Pkinesis-asl clean package
```
@JoshRosen Could you add `-Psparkr` to that and other QA builds ?

**Pull request closed without merging**